### ISSUE TITLE
clamav 0.104.0 libcursesw support

### ISSUE THREAD

**Issue by dilyanpalauzov** [Cisco-Talos/clamav#292]
I do not need the curses apps.  clamav 0.103.3/configure worked without curses, the Cmake build fails, if it cannot find libcurses.  For the record, I have instead libncursesw .

* Permit building clamd, freshclam, milter, and others, when curses are missing.

**Comment by dilyanpalauzov** [comment]
So, I want to compile clamd and clamscan, without clamdtop, but the build system does not allow me)

**Comment by micahsnyder** [comment]
I've been thinking about how to respond to this. Yes clamdtop is not optional when configuring the build in 0.104.  Making it optional like it was with the previous build system would be mostly trivial. But reducing build configuration complexity was one of my goals when starting over with CMake, which is why the option doesn't exist. I don't want to have to test additional build permutations to serve a very small minority that could get by simply by installing all of the project dependencies. This is why I'm hesitant to bring it back.  I'm inclined to say "sorry, you need ncurses or pdcurses." Mixed feelings. 🙁

But you raise the point that you have ncursesw.  I hadn't heard of ncursesw before.  I did a search and found this helpful page describing the library and common pitfalls when adding support for it http://www.roguebasin.com/index.php?title=Ncursesw
So I guess ncursesw is a fork of ncurses to support UTF-8 that "is surprisingly difficult to get working with C programs". 
I do have a desire to improve ClamAV's UTF-8 support.  So ... now I am wondering if we should try to support ncursesw in the future?

For now I guess I need to know: does ncurses play nicely with ncursesw?  Can you install ncurses to get over this hurdle and we can deal with ncursesw some other day?

**Comment by dilyanpalauzov** [comment]
I run Linux from Scratch, I have setup ncursesw and I do not want to change it, as this can cause problems with booting/logging in, since bash/readline also uses ncursesw.

ncurses and ncursesw are compiled from the same source code from https://invisible-island.net/ncurses/ncurses.html . I see now only ftp links for downloading tagged relleases. I compile it with

> ./configure --enable-widec --without-cxx-binding --enable-pc-files --with-shared --without-debug --without-normal --with-termlib --with-ticlib --enable-sp-funcs --enable-ext-colors --enable-ext-mouse --enable-ext-putwin --enable-sigwinch

I guess, --without-normal avoids creating libncurses while --enable-widec creates libncursesw .

**Comment by adb014** [comment]
Isn't just as simple as ?

```
--- cmake/FindCURSES.cmake.old	2021-12-07 03:35:40.902075699 +0000
+++ cmake/FindCURSES.cmake	2021-12-07 03:43:38.172042783 +0000
@@ -134,7 +134,52 @@
         )
       endif()
   else()
-      message(FATAL_ERROR "Unable to find ncurses or pdcurses")
+    # Try for NCursesw
+    pkg_check_modules(PC_NCursesw QUIET ncursesw)
+
+    find_path(NCURSESW_INCLUDE_DIR
+      NAMES ncurses.h
+      PATHS \${PC_NCursesw_INCLUDE_DIRS} \${CURSES_INCLUDE_DIR}
+    )
+
+    string(FIND \${NCURSESW_INCLUDE_DIR} "-NOTFOUND" NCURSESW_NOT_FOUND)
+    if(NCURSESW_NOT_FOUND EQUAL -1)
+      #
+      # ncursesw WAS found!
+      #
+      set(HAVE_LIBNCURSESW 1)
+      set(CURSES_INCLUDE "<ncurses.h>")
+
+      set(CURSES_LIBRARY \${PC_NCursesw_LINK_LIBRARIES})
+      set(CURSES_VERSION \${PC_NCursesw_VERSION})
+
+      include(FindPackageHandleStandardArgs)
+      find_package_handle_standard_args(CURSES
+        FOUND_VAR CURSES_FOUND
+        REQUIRED_VARS
+          CURSES_LIBRARY
+          NCURSESW_INCLUDE_DIR
+        VERSION_VAR CURSES_VERSION
+      )
+
+      set(HAVE_LIBNCURSESW 1)
+      set(CURSES_INCLUDE "<ncurses.h>")
+
+      set(CURSES_LIBRARIES \${CURSES_LIBRARY})
+      set(CURSES_INCLUDE_DIRS \${NCURSESW_INCLUDE_DIR})
+      set(CURSES_DEFINITIONS \${PC_NCursesw_CFLAGS_OTHER})
+
+      if (NOT TARGET Curses::curses)
+        add_library(Curses::curses INTERFACE IMPORTED)
+        set_target_properties(Curses::curses PROPERTIES
+          INTERFACE_COMPILE_OPTIONS "\${PC_NCursesw_CFLAGS_OTHER}"
+          INTERFACE_INCLUDE_DIRECTORIES "\${CURSES_INCLUDE_DIRS}"
+          INTERFACE_LINK_LIBRARIES "\${CURSES_LIBRARY}"
+        )
+      endif()
+    else()
+        message(FATAL_ERROR "Unable to find ncurses, ncursesw or pdcurses")
+    endif()
   endif()
 endif()
```

**Comment by micahsnyder** [comment]
This issue should be resolved by:
https://github.com/Cisco-Talos/clamav/pull/355
https://github.com/Cisco-Talos/clamav/pull/364

The fixes have also been backported for the 0.104.3 patch release here: 
https://github.com/Cisco-Talos/clamav/pull/480
